### ISSUE TITLE
The access token expiry UTC time '6/29/2018 8:53:02 PM' is earlier than current UTC time '7/2/2018 9:46:18 PM'.

### ISSUE THREAD

**Issue by vladimir-shcherbakov** [Azure/azure-powershell#6585]
### Description

After successful login *AzureRm* cmdlets throw the excetpion : " The access token expiry UTC time  '6/29/2018 8:53:02 PM' is earlier than current UTC time '7/2/2018 9:46:18 PM'."

### Script/Steps for Reproduction

```powershell
PS C:\> Login-AzureRmAccount


Account          : <hidden>@microsoft.com
SubscriptionName : <hidden>
SubscriptionId   : <hidden>
TenantId         : 72f988bf-86f1-41af-91ab-2d7cd011db47
Environment      : AzureCloud

PS C:\> Get-AzureRmSubscription
WARNING: Unable to acquire token for tenant '72f988bf-86f1-41af-91ab-2d7cd011db47'
Get-AzureRmSubscription : The access token expiry UTC time '6/29/2018 8:53:02 PM' is earlier than current UTC time
'7/2/2018 9:46:18 PM'.
At line:1 char:1
+ Get-AzureRmSubscription
+ ~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : CloseError: (:) [Get-AzureRmSubscription], CloudException
    + FullyQualifiedErrorId : Microsoft.Azure.Commands.Profile.GetAzureRMSubscriptionCommand
```

### Module Version

```powershell
PS C:\> Get-Module -Name AzureRM -ListAvailable


    Directory: C:\Program Files\WindowsPowerShell\Modules


ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     6.3.0      AzureRM
```

### Environment Data

```powershell
PS C:\> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      5.1.17134.112
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.17134.112
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

### Debug Output

```

```

**Comment by cormacpayne** [comment]
@vladimir-shcherbakov are you seeing this issue after restarting your PowerShell session? If so, run `Clear-AzureRmContext`, open a new session and login again using `Connect-AzureRmAccount`. If you still see the same error, set $DebugPreference="Continue" and return the result of the cmdlet execution with the debug stream.

**Comment by vladimir-shcherbakov** [comment]
I opened a new PS terminal and logged in.
Of course, Clear-AzureRmContex solves such an issues. 
I added the issue as an yet another sample of token cache related issues.

**Comment by edyoung** [comment]
Likely related to #6328

**Comment by dhmorrison** [comment]
I can repo this in the _Azure Portal Cloud Shell:_

1. Execute a command without parameters
1. Provide input values as the system prompts for _required parameters_
1. Input a value that is not to the liking of the shell, perhaps it wants a single quoted parameter
1. Profit!  I"M _HORDING_ all the single quotes for my own use.  lol

```
cmdlet New-AzResourceGroupDeployment at command pipeline position 1
Supply values for the following parameters:
(Type !? for Help.)
ResourceGroupName: sqlcloneprodrgrp
TemplateFile: azuredeploy.json
```

**BLAM!**  _Errors will be happening_ as more than one contractor has reported to me...
It would be super awesome if we can eliminate this, as I think we all get the promise of Azure resource group deployments.  IMO, it's pretty challenging to get them debugged and sorted out in the best circumstances.  Errors (false errors-can an error be true || false?) kind of add insult to injury.  Disrupts whatever little bit of flow I'm eking out.
______________
```
cmdlet New-AzResourceGroupDeployment at command pipeline position 1
Supply values for the following parameters:
(Type !? for Help.)
ResourceGroupName: sqlcloneprodrgrp
TemplateFile: azuredeploy.json
New-AzureRmResourceGroupDeployment : The access token expiry UTC time '1/21/2019 11:27:25 PM' is earlier than current UTC time '1/21/2019 11:32:17 PM'
At line:1 char:1
+ New-AzureRmResourceGroupDeployment
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : CloseError: (:) [New-AzResourceGroupDeployment], CloudException
+ FullyQualifiedErrorId : Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation.NewAzureResourceGroupDeploymentCmdlet
```
______________

**Comment by nycbobby** [comment]
Please see the following link: https://docs.microsoft.com/en-us/azure/cloud-shell/limitations

It states the following: "Cloud Shell is intended for interactive use cases. As a result, any long-running non-interactive sessions are ended without warning."

This is not very intuitive considering Cloud Shell has an editor to write & save long script files which one would assume you could use for Azure VM deployments. I've had problems with this token expiring at random times. Sometimes 40 minutes after restarting Cloud Shell, sometimes 18 minutes. 

The solution is to use Powershell from local PC and sign-in to Azure through the script-pane. Please see: https://www.jgspiers.com/how-to-connect-to-azure-powershell-arm-azuread

**Comment by slavizh** [comment]
@cormacpayne is there any workaround for this issue? I have tried using connect-AzAccount while still executing new-AzDeploymnet -verbose but seems the command is using the same access token and does not get the latest from the Az context. Either the commands should always get the latest access token to avoid expiration during waiting for results from commands or may be it is good to implement login via connect-AzAccount with refresh token instead of access token so you can renew it.

**Comment by cormacpayne** [comment]
@slavizh Hey Stanislav, I'm no longer working on the Azure PowerShell team as of a few months ago -- @dingmeng-xue @erich-wang @isra-fel would one of you mind taking a look at this issue?

**Comment by dingmeng-xue** [comment]
@slavizh , could you set `$DebugPreference="Continue"` and try `Get-AzSubscription`? Then please share your output to us.

**Comment by slavizh** [comment]
@dingmeng-xue I will but let me explain what I am doing exactly. So I am executing this in DevOps pipeline. The main logic is the following:
- I get access token by using refresh token. The access token is good for 1 hour. I cannot find any way to extend it even with Azure AD policies. Seems they do not apply when the resource is https://management.azure.com
- I login with Az PowerShell to Azure with that token
- After I login I also start a background job with Start-ThreadJob. This runs in the same session as the current one I am in. That background job does the following: when the access token is about to expire in 5 minutes I issue new access token with new expire date and I use connect-azaccount to login with -Force switch. By doing this when new az commands are executed they will use the new token.
- While the background job is running but hasn't issued a new token and re-login I start arm template deployment with New-AzDeployment and -Verbose switch. The deployment starts. This deployment takes more than 1 hour so when that hour passes  New-AzDeployment  throws the above error. The background job has already renewed the token and re-login but New-AzDeployment while it was running did not recognize that and was using the old access token still. If I run other Az commands after New-AzDeployment has thrown that error they run fine because they use the new access token.

**Comment by slavizh** [comment]
@dingmeng-xue if you can give e-mail address I can send you the debug information on e-mail. There is sensitive information in the debug that I do not want to expose here.

**Comment by dingmeng-xue** [comment]
@slavizh , your instruction is very useful. We will check the code and analyze potential reason. If we need further information, we will ping you.

**Comment by slavizh** [comment]
Thanks!
The above problem I think it is similar but the initial one I have posted is more important. In case it might help I am posting it here.
I was executing this:

```
$allDeployments = Get-AzResourceGroupDeployment -ResourceGroupName $resourceGroupName -ErrorAction Stop
$last100Deployments = $allDeployments | Sort-Object -Property Timestamp | Select-Object -First 100
$last100Deployments | Remove-AzResourceGroupDeployment -ErrorAction Continue
```

In that case I received some error message of being 'unauthorized'. Deleting deployments takes time so I have assume that it took more than one hour. It deleted  the bigger part of the deployments but several gave the unauthorized error. Although I am renewing the access token and re-connecting I am not sure why when the code is executed this way not all commands took the latest access token.
In any case the problem with New-AzDeployment is more important than this one even if connected.

**Comment by slavizh** [comment]
Alternative option would be to allow login with Refresh token and you can issue access tokens for Azure management, graph and key vault and renew them before they expire on the background.

**Comment by slavizh** [comment]
@dingmeng-xue Is there any update on this? This problem is critical for us.

**Comment by dingmeng-xue** [comment]
@slavizh , I couldn't reproduce your problem because I haven't figure out how to shorten token expiration date in my environment.  When I went through the code, I found long running operation job in resource management SDK starts a new thread and I suspect it leads concurrency issue. I will drop you an email and do further investigation.

**Comment by dingmeng-xue** [comment]
Update about issue. The root cause is access token used by long running operation of deployment may be expired. There is no way to set refresh token for the time being or allow executing cmdlet to get new access token. We need to allow access token be updated for long running operation.

**Comment by dingmeng-xue** [comment]
This one may be mitigated by #11925 and #11924

**Comment by VeryEarly** [comment]
fixed in #12454
